### ISSUE TITLE
Better redis compliancy for the inmemory stub.

### ISSUE THREAD

**Issue by GlenDC** [g8os/blockstor#46]
Currently the in-memory Redis stub client doesn't support pipelining.

+ `MULTI` and `EXEC` commands are recognized but ignored;
+ `SEND` only supports a small subset of the supported `Do` commands;
+ `SEND` does not collect its results for later use (eg. when `EXEC` gets executed);
+ `Flush` doesn't do anything either for now;

Support for this would be greatly welcome for the WIP code of https://github.com/g8os/blockstor/pull/44.

**EDIT**:

TODO:

- [x] Create InMemoryRedis Conn, using [LedisDB](https://github.com/siddontang/ledisdb);
  - [x] Support Lua Scripting (`EVAL` command);
  - [x] Support pipelining (`Send`, `Flush`, `MULTI`, `EXEC` (fake it));
- [x] Support `redis` global variable _ledisdb_'s Lua support (as an alias of `ledis` if possible), see: https://github.com/siddontang/ledisdb/pull/289 (merged);
- [x] Replace [CGO Lua](https://github.com/siddontang/golua) (in LedisDB) with [Gopher Lua](https://github.com/yuin/gopher-lua) (see: https://github.com/siddontang/ledisdb/issues/288) (PR: https://github.com/siddontang/ledisdb/pull/291);
- [x] Allocate a port dynamically for ledisdb, when no address is specified (see: https://github.com/siddontang/ledisdb/issues/290) (PR: https://github.com/siddontang/ledisdb/pull/292);
- [ ] Improve InMemoryRedis performance, as it seems to account for most of the cpu usage when used (ledisdb, and especially `net(*conn).Write/Read`, seem responsible) (solution proposed in open PR);
- [x] Vendor dependencies;

**Comment by robvanmieghem** [comment]
As suggested earlier, it might be better to embed a real redis implementation and reuse the normal redigo client. http://ledisdb.com might be an option to embed.

**Comment by GlenDC** [comment]
Yeah I agree, first I thought it was quite the work, but do-able, to make it feature-complete. However I forgot about the `EVAL`, `EVALSHA` commands, until I started adding Lua scripts myself to a feature branch of this repo. And then I realized that's it's pretty insane to do it ourselves.

LedisDB looks interesting. Want me to do some experimenting with it @robvanmieghem? Can do it today, and benchmark to see if it's worth it. Or do you have some data on it already? On the benchmarks they provide, it seems to provide a _bit_ worse than Redis :|

**Comment by iwanbk** [comment]
```
  Can do it today, and benchmark to see if it's worth it. Or do you have some data on it already?
```
We should have some benchmark data between ledisdb & ardb, because we use ledisdb in previous qemu vol driver.
```
On the benchmarks they provide, it seems to provide a bit worse than Redis :
```
yep, it store the data on disk.

**Comment by robvanmieghem** [comment]
@GlenDC Yes, I would be good to to embed it. No reason to benchmark though, the embedded redis would only be for development and testing, not for performance.

**Comment by robvanmieghem** [comment]
@GlenDC regarding "Create InMemoryRedis Conn, , using LedisDB;", can't we just start a ledisdb and use the normal redigo client to connect to it?

**Comment by GlenDC** [comment]
@robvanmieghem after reading through the ledisdb server code I can confirm that you are correct, and after quickly hacking it in nbdserver I can double confirm that you are correct. Sorry for that, should have read more thoroughly the ledisdb code yesterday. Will update my task, and after breakfast I'll turn my quick hack into a clean solution. Will then push it as WIP PR, after which I'll integrate gopherlua into a forked ledisdb repo. We'll use that repo for now, as the maintainer of ledisdb seems to be very busy now, as he told me, so I don't see this being merged in the next 1 to 2 weeks. Once it's merged we can than use the original ledisdb, once again.

**Comment by GlenDC** [comment]
Another update. We might not need the `MULTI` / `EXEC` functionality after all. So for now I'm going to try to do it without the stub, and try to keep it with as pure ledisdb code as possible. We'll still have to solve the `ledis` instead of `redis` global lua variable, so let's see what we can do about that.

**Comment by GlenDC** [comment]
Yet Another Update. The `ledis`/`redis` issue is resolved, as I've merged into ledisdb a fix for it. So now `redis` is accepted as well. With that being confirmed and merged, I've started focussing on merging in gopher-lua. This will allow us to use Lua without having to rely on CGO, and instead stay 100% pure Go. I've started on this, and it's for the most part done, as all code has been converted within ledisdb and some unit-tests already pass. But it was a relatively big, and certainly invasive, commit. There are a few bugs ATM. So will have to fix those, and test very well. Once it's all working and confirmed, I'll open a PR in ledisdb.

My previous (and first) PR, that I requested today, was merged in within a couple of hours, so fingers crossed that the next PR will be resolved as fast.

**Comment by robvanmieghem** [comment]
We do have a fixed port for now and it would be better to have one dynamically allocated, this does require a change in ledisdb: https://github.com/siddontang/ledisdb/issues/290

**Comment by GlenDC** [comment]
Good idea, will add it to do the tasks of this issue!

**Comment by GlenDC** [comment]
This issue is mostly resolved in the open PR (#51). The only open task left is improving the performance of the embedded ledisdb (memorystorage). A solution for which has been proposed in the open PR (#51).
